services/motion/explore/explore.go

```diff
@@ -0,0 +1,575 @@
+	}, ms.backgroundWorkers.Done)
+
+	// Start executing plan
+	ms.backgroundWorkers.Add(1)
+	goutils.ManagedGo(func() {
+		ms.executePlan(cancelCtx, kb, plan)
+	}, ms.backgroundWorkers.Done)
+
+	for {
+		// this ensures that if the context is cancelled we always return early at the top of the loop
+		if err := ctx.Err(); err != nil {
+			return false, err
+		}
+
+		select {
+		// if context was cancelled by the calling function, error out
+		case <-ctx.Done():
+			return false, ctx.Err()
+
+		// once execution responds: return the result to the caller
+		case resp := <-ms.executionResponseChan:
+			ms.logger.Debugf("execution completed: %s", resp)
+			return resp.success, resp.err
+
+		// if the checkPartialPlan process hit an error return it, otherwise exit
+		case resp := <-ms.obstacleResponseChan:
+			ms.logger.Debugf("obstacle response: %s", resp)
+			if resp.err != nil {
+				return resp.success, resp.err
+			}
+			if resp.success {
+				return resp.success, nil
+			}
+		}
+	}
+}
+
+type moveResponse struct {
+	err     error
+	success bool
+}
+
+// checkForObstacles will continuously monitor the generated transient worldState for obstacles in the given
+// motionplan plan. A response will be sent through the channel if an error occurs, the motionplan plan
+// completes or an obstacle is detected in the given range.
+func (ms *explore) checkForObstacles(
+	ctx context.Context,
+	obstacleDetectors []obstacleDetectorObject,
+	kb kinematicbase.KinematicBase,
+	plan motionplan.Plan,
+	obstaclePollingFrequencyHz float64,
+) {
+	// Constantly check for obstacles in path at desired obstacle polling frequency
+	ticker := time.NewTicker(time.Duration(int(1000/obstaclePollingFrequencyHz)) * time.Millisecond)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			currentPose := spatialmath.NewZeroPose()
+
+			// Look for new transient obstacles and add to worldState
+			worldState, err := ms.generateTransientWorldState(ctx, obstacleDetectors)
+			if err != nil {
+				ms.obstacleResponseChan <- moveResponse{err: err}
+				return
+			}
+
+			// Check motionplan plan for transient obstacles
+			collisionPose, err := motionplan.CheckPlan(
+				kb.Kinematics(),
+				plan,
+				worldState,
+				ms.frameSystem,
+				currentPose,
+				[]referenceframe.Input{{Value: 0}, {Value: 0}},
+				spatialmath.NewZeroPose(),
+				ms.logger,
+			)
+			if err != nil {
+				// If an obstacle is detected, check if its within the valid obstacle distance to trigger an
+				// end to the checkForObstacle loop
+				if collisionPose.Point().Distance(currentPose.Point()) < validObstacleDistanceMM {
+					ms.logger.Debug("collision found")
+					ms.obstacleResponseChan <- moveResponse{success: true, err: err}
+					return
+				}
+				ms.logger.Debug("collision found but outside of range")
+				ms.obstacleResponseChan <- moveResponse{success: false, err: err}
```

Comment:
I purposefully left all these channel responses in because I wished to discuss when we would like to send a message back to the main system. We can either:

Report back to channel with any and all information, casing when to exit based on the moveResponse
Error with worldState: `moveResponse{err: err}`
Collision detected in valid region: `moveResponse{success: true, err: err}`
Collision detected but out of range: `moveResponse{success: false, err: err}`
Collision not detected: `moveResponse{success: false, err: err}`

```
case resp := <-ms.obstacleResponseChan:
	ms.logger.Debugf("obstacle response: %s", resp)
	if resp.err != nil && !resp.err.Error().Contains("collision detected){
		return false, resp.err
	}
	if resp.success {
		return true, nil
	}
```

OR


Only send a moveResponse to the channel if a collision was detected in the valid region
Error with worldState: `ignore or return (not sending info to channel)`
Collision detected in valid region: `moveResponse{success: true, err: err}`
Collision detected but out of range: `ignore`
Collision not detected: `ignore`

```
case resp := <-ms.obstacleResponseChan:
	ms.logger.Debugf("obstacle response: %s", resp)
	return resp.success, resp.err
```

Note #1: In this second case, the channel message type `moveResponse` could be replaced with a `bool` channel.
Note #2: If we ignore the worldState check then we may enter a situation where the worldState fails to build every iteration and we continue moving indefinitely. If we report the error, then a failure to build worldState could/would result in a new point being chosen to head towards. Basically it comes down to if we want to ignore errors and continue to move in the desired direction until a timeout is reached OR change destination locations as frequently as worldState fails to build.

I am in favor of option 2, however, I was hoping the motion team (@biotinker, @nfranczak, @raybjork) could comment on their choices of when to return an error in their builtIn representation of MoveOnMap which uses a similar channel/goroutine process.

Reply:
I think checkPlan should only write to obstacleResponseChan when it wants to terminate the Move call (because there is an obstacle within range). But I don't think I understand when/why the worldState would fail to build.

Reply:
The only two way creating the world state could fail is if GetObjectPointClouds fails or if NewWorldState() fails. The first of these may occur. The second should never. 

For GetObjectPointClouds failing, we are likely fine to ignore it as it is unlikely to be a persistent problem.

Reply:
Got it. For MoveOnGlobe, if GetObjectPointClouds fails, it returns an error, and the navigation service logs the error and retries. I think we can do the same thing here (where I don't have a preference between retrying the same point and retrying a new random point).